spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/SparkParquetReaders.java

```diff
@@ -187,13 +187,14 @@ private SparkParquetReaders() {
     @Override
     public ParquetValueReader<?> list(Types.ListType expectedList, GroupType array,
                                       ParquetValueReader<?> elementReader) {
-      GroupType repeated = array.getFields().get(0).asGroupType();
+      Type repeated = array.getFields().get(0);
+      boolean isOldListElementType = ParquetSchemaUtil.isOldListElementType(repeated, array.getName());
       String[] repeatedPath = currentPath();
 
       int repeatedD = type.getMaxDefinitionLevel(repeatedPath) - 1;
       int repeatedR = type.getMaxRepetitionLevel(repeatedPath) - 1;
 
-      Type elementType = repeated.getType(0);
+      Type elementType = isOldListElementType ? repeated : repeated.asGroupType().getType(0);
```

Comment:
Looks like this doesn't fix Flink or Generics?

Reply:
Added. As we discussed, I will backport to older versions of Spark and flink in a separate diff.